File: pageindicatorview/src/main/java/com/rd/draw/drawer/type/WormDrawer.java
package com.rd.draw.drawer.type;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import androidx.annotation.NonNull;
import com.rd.animation.data.Value;
import com.rd.animation.data.type.WormAnimationValue;
import com.rd.draw.data.Indicator;
import com.rd.draw.data.Orientation;

public class WormDrawer extends BaseDrawer {

    public RectF rect;

    public WormDrawer(@NonNull Paint paint, @NonNull Indicator indicator) {
        super(paint, indicator);
        rect = new RectF();
    }

    public void draw(
            @NonNull Canvas canvas,
            @NonNull Value value,
            int coordinateX,
            int coordinateY) {

        if (!(value instanceof WormAnimationValue)) {
            return;
        }

        WormAnimationValue v = (WormAnimationValue) value;
        int rectStart = v.getRectStart();
        int rectEnd = v.getRectEnd();

        int radius = indicator.getRadius();
        int unselectedColor = indicator.getUnselectedColor();
        int selectedColor = indicator.getSelectedColor();

        if (indicator.getOrientation() == Orientation.HORIZONTAL) {
            rect.left = rectStart;
            rect.right = rectEnd;
            rect.top = coordinateY - radius;
            rect.bottom = coordinateY + radius;

        } else {
            rect.left = coordinateX - radius;
            rect.right = coordinateX + radius;
            rect.top = rectStart;
            rect.bottom = rectEnd;
        }

        paint.setColor(unselectedColor);
        canvas.drawCircle(coordinateX, coordinateY, radius, paint);

        paint.setColor(selectedColor);
        canvas.drawRoundRect(rect, radius, radius, paint);
    }
}

File: pageindicatorview/src/main/java/com/rd/draw/drawer/type/WormDrawer.kt
package com.rd.draw.drawer.type

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import com.rd.animation.data.Value
import com.rd.animation.data.type.WormAnimationValue
import com.rd.draw.data.Indicator
import com.rd.draw.data.Orientation

open class WormDrawer(paint: Paint, indicator: Indicator) : BaseDrawer(paint, indicator) {

	var rect: RectF = RectF()

	open fun draw(
			canvas: Canvas,
			value: Value,
			coordinateX: Int,
			coordinateY: Int) {

		if (value !is WormAnimationValue) {
			return
		}

		val rectStart = value.rectStart
		val rectEnd = value.rectEnd

		val radius = indicator.radius
		val unselectedColor = indicator.unselectedColor
		val selectedColor = indicator.selectedColor

		if (indicator.orientation === Orientation.HORIZONTAL) {
			rect.left = rectStart.toFloat()
			rect.right = rectEnd.toFloat()
			rect.top = (coordinateY - radius).toFloat()
			rect.bottom = (coordinateY + radius).toFloat()

		} else {
			rect.left = (coordinateX - radius).toFloat()
			rect.right = (coordinateX + radius).toFloat()
			rect.top = rectStart.toFloat()
			rect.bottom = rectEnd.toFloat()
		}

		paint.color = unselectedColor
		canvas.drawCircle(coordinateX.toFloat(), coordinateY.toFloat(), radius.toFloat(), paint)

		paint.color = selectedColor
		canvas.drawRoundRect(rect, radius.toFloat(), radius.toFloat(), paint)
	}
}
